wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,238 @@
+   *                               changes. The value specified should be in nanoseconds.
+   * @param rdsClientFunc          The function to call to create the RDS client that will fetch custom endpoint
+   *                               information.
+   */
+  public CustomEndpointMonitorImpl(
+      PluginService pluginService,
+      HostSpec customEndpointHostSpec,
+      Region region,
+      long refreshRateNano,
+      BiFunction<HostSpec, Region, RdsClient> rdsClientFunc) {
+    this.pluginService = pluginService;
+    this.customEndpointHostSpec = customEndpointHostSpec;
+    this.region = region;
+    this.refreshRateNano = refreshRateNano;
+    this.rdsClient = rdsClientFunc.apply(customEndpointHostSpec, this.region);
+
+    this.endpointIdentifier = this.rdsUtils.getRdsClusterId(customEndpointHostSpec.getHost());
+    if (StringUtils.isNullOrEmpty(this.endpointIdentifier)) {
+      throw new RuntimeException(
+          Messages.get(
+              "CustomEndpointMonitorImpl.errorParsingEndpointIdentifier",
+              new Object[] {customEndpointHostSpec.getHost()}));
+    }
+
+    TelemetryFactory telemetryFactory = this.pluginService.getTelemetryFactory();
+    this.infoChangedCounter = telemetryFactory.createCounter(TELEMETRY_ENDPOINT_INFO_CHANGED);
+
+    this.monitorExecutor.submit(this);
+    this.monitorExecutor.shutdown();
+  }
+
+  /**
+   * Analyzes a given custom endpoint for changes to custom endpoint information.
+   */
+  @Override
+  public void run() {
+    LOGGER.fine(
+        Messages.get(
+            "CustomEndpointMonitorImpl.startingMonitor",
+            new Object[] { this.customEndpointHostSpec.getHost() }));
+
+    try {
+      while (!this.stop.get() && !Thread.currentThread().isInterrupted()) {
+        long start = System.nanoTime();
+
+        final Filter customEndpointFilter =
+            Filter.builder().name("db-cluster-endpoint-type").values("custom").build();
+        final DescribeDbClusterEndpointsResponse endpointsResponse =
+            this.rdsClient.describeDBClusterEndpoints(
+                (builder) ->
+                    builder.dbClusterEndpointIdentifier(this.endpointIdentifier).filters(customEndpointFilter));
+
+        List<DBClusterEndpoint> endpoints = endpointsResponse.dbClusterEndpoints();
+        if (endpoints.size() != 1) {
+          List<String> endpointURLs = endpoints.stream().map(DBClusterEndpoint::endpoint).collect(Collectors.toList());
+          LOGGER.warning(
+              Messages.get("CustomEndpointMonitorImpl.unexpectedNumberOfEndpoints",
+                  new Object[] {
+                      this.endpointIdentifier,
+                      this.region.id(),
+                      endpoints.size(),
+                      endpointURLs
+                  }
+              ));
+
+          TimeUnit.NANOSECONDS.sleep(this.refreshRateNano);
+          continue;
+        }
```

Comment:
Should we report about internal issue if the list contains more than one element?

Reply:
Its definitely an exceptional case, which is why I log a warning about this above this line. I would normally throw a SQLException if this happened in the main thread, but this is happening in a monitor thread, so I'm not sure how to report the exception to the main thread.